src/algebra/periodic.lean

```diff
@@ -0,0 +1,385 @@
+/-
+Copyright (c) 2021 Benjamin Davidson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Benjamin Davidson
+-/
+import data.int.parity
+import algebra.module
+
+/-!
+# Periodicity
+
+In this file we define and then prove facts about periodic and antiperiodic functions.
+
+## Main definitions
+
+* `function.periodic`: A function `f` is *periodic* if `∀ x, f (x + c) = f x`.
+  `f` is referred to as periodic with period `c` or `c`-periodic.
+
+* `function.antiperiodic`: A function `f` is *antiperiodic* if `∀ x, f (x + c) = -f x`.
+  `f` is referred to as antiperiodic with antiperiod `c` or `c`-antiperiodic.
+
+Note that any `c`-antiperiodic function will necessarily also be `2*c`-periodic.
+
+## Tags
+
+period, periodic, periodicity, antiperiodic
+-/
+
+variables {α β γ : Type*} {f g : α → β} {c c₁ c₂ x : α}
+
+namespace function
+
+/-! ### Periodicity -/
+
+/-- A function `f` is said to be `periodic` with period `c` if for all `x`, `f (x + c) = f x`. -/
+@[simp] def periodic [has_add α] (f : α → β) (c : α) : Prop :=
+∀ x : α, f (x + c) = f x
+
+lemma periodic.comp [has_add α]
+  (h : periodic f c) (g : β → γ) :
+  periodic (g ∘ f) c :=
+by simp * at *
+
+@[to_additive]
+lemma periodic.mul [has_add α] [has_mul β]
+  (hf : periodic f c) (hg : periodic g c) :
+  periodic (f * g) c :=
+by simp * at *
+
+@[to_additive]
+lemma periodic.div [has_add α] [has_div β]
+  (hf : periodic f c) (hg : periodic g c) :
+  periodic (f / g) c :=
+by simp * at *
+
+lemma periodic.const_smul [add_monoid α] [group γ] [distrib_mul_action γ α]
+  (h : periodic f c) (a : γ) :
+  periodic (λ x, f (a • x)) (a⁻¹ • c) :=
+λ x, by simpa only [smul_add, smul_inv_smul] using h (a • x)
+
+lemma periodic.const_smul' [add_comm_monoid α] [division_ring γ] [module γ α]
+  (h : periodic f c) (a : γ) :
+  periodic (λ x, f (a • x)) (a⁻¹ • c) :=
+begin
+  intro x,
+  by_cases ha : a = 0, { simp only [ha, zero_smul] },
+  simpa only [smul_add, smul_inv_smul' ha] using h (a • x),
+end
+
+lemma periodic.const_mul [division_ring α]
+  (h : periodic f c) (a : α) :
+  periodic (λ x, f (a * x)) (a⁻¹ * c) :=
+h.const_smul' a
+
+lemma periodic.const_inv_smul [add_monoid α] [group γ] [distrib_mul_action γ α]
+  (h : periodic f c) (a : γ) :
+  periodic (λ x, f (a⁻¹ • x)) (a • c) :=
+by simpa only [inv_inv] using h.const_smul a⁻¹
+
+lemma periodic.const_inv_smul' [add_comm_monoid α] [division_ring γ] [module γ α]
+  (h : periodic f c) (a : γ) :
+  periodic (λ x, f (a⁻¹ • x)) (a • c) :=
+by simpa only [inv_inv'] using h.const_smul' a⁻¹
+
+lemma periodic.const_inv_mul [division_ring α]
+  (h : periodic f c) (a : α) :
+  periodic (λ x, f (a⁻¹ * x)) (a * c) :=
+h.const_inv_smul' a
+
+lemma periodic.mul_const [division_ring α]
+  (h : periodic f c) (a : α) :
+  periodic (λ x, f (x * a)) (c * a⁻¹) :=
+begin
+  intro x,
+  by_cases ha : a = 0, { simp only [ha, mul_zero] },
+  simpa only [add_mul, inv_mul_cancel_right' ha] using h (x * a),
+end
```

Comment:
It probably needs a pre-work PR, but I can reuse the proof if I define:
```lean
instance monoid.opposite_to_mul_action (R : Type*) [monoid R] : mul_action (opposite R) R :=
{ smul := λ c x, x * c.unop,
  one_smul := by simp,
  mul_smul := λ r s x, by simp [mul_assoc] }

instance semiring.opposite_to_module (R : Type*) [semiring R] : module (opposite R) R :=
{ smul := λ c x, x * c.unop,
  smul_add := sorry,
  smul_zero := sorry,
  zero_smul := sorry,
  add_smul := sorry,
  .. monoid.opposite_to_mul_action R }

instance (R : Type*) [division_ring R] : division_ring (opposite R) :=
{ inv := λ x, opposite.op (x.unop⁻¹),
  mul_inv_cancel := sorry,
  inv_zero := sorry,
  ..opposite.ring R, ..opposite.nontrivial R}

lemma periodic.mul_const [division_ring α]
  (h : periodic f c) (a : α) :
  periodic (λ x, f (x * a)) (c * a⁻¹) :=
h.const_smul' (opposite.op a)
```
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Right.20multiplication.20as.20a.20mul_action/near/239012156)

Reply:
I've PR'd the first two instances in #7630.

Reply:
Which is now merged

Reply:
Thank you, Eric!

Reply:
(Now merged)

Reply:
This needs one more preliminary PR. See # 7698.